test/metabase/notification/send_test.clj

```diff
@@ -120,13 +124,218 @@
+          (t2/delete! :model/TaskHistory :task "channel-send")
+          (testing "and record exception in task history"
+            (let [send!       (fn [& _args]
+                                (throw (ex-info "Failed to send" {:metadata 42})))]
+              (mt/with-dynamic-fn-redefs [notification.send/should-skip-retry? (constantly true)
+                                          channel/send!                        send!]
+                (#'notification.send/send-notification-sync! n))
+              (is (=? {:task "channel-send"
+                       :status       :failed
+                       :task_details {:attempted_retries 1
+                                      :message           "Failed to send"
+                                      :ex-data           {:metadata 42
+                                                          :metabase.notification.send/skip-retry? true}
+
+                                      :retry_errors      (mt/malli=? [:sequential [:map {:closed true}
+                                                                                   [:timestamp :string]
+                                                                                   [:message :string]]])}}
+                      (t2/select-one :model/TaskHistory :task "channel-send"))))))))))
+
+(defn- get-positive-retry-metrics [^io.github.resilience4j.retry.Retry retry]
+  (let [metrics (bean (.getMetrics retry))]
+    (into {}
+          (map (fn [field]
+                 (let [n (metrics field)]
+                   (when (pos? n)
+                     [field n]))))
+          [:numberOfFailedCallsWithRetryAttempt
+           :numberOfFailedCallsWithoutRetryAttempt
+           :numberOfSuccessfulCallsWithRetryAttempt
+           :numberOfSuccessfulCallsWithoutRetryAttempt])))
+
+(def ^:private fake-email-notification
+  {:subject      "test-message"
+   :recipients   ["whoever@example.com"]
+   :message-type :text
+   :message      "test message body"})
+
+(def ^:private test-retry-configuration
+  (assoc @#'notification.send/default-retry-config
+         :initial-interval-millis 1
+         :max-attempts 2))
+
+(deftest email-notification-retry-test
+  (testing "send email succeeds w/o retry"
+    (let [test-retry (retry/random-exponential-backoff-retry "test-retry" test-retry-configuration)]
+      (with-redefs [email/send-email!                      mt/fake-inbox-email-fn
+                    retry/random-exponential-backoff-retry (constantly test-retry)]
+        (mt/with-temporary-setting-values [email-smtp-host "fake_smtp_host"
+                                           email-smtp-port 587]
+          (mt/reset-inbox!)
+          (#'notification.send/channel-send-retrying! 1 :notification/card {:channel_type :channel/email} fake-email-notification)
+          (is (= {:numberOfSuccessfulCallsWithoutRetryAttempt 1}
+                 (get-positive-retry-metrics test-retry)))
+          (is (= 1 (count @mt/inbox)))))))
```

Comment:
We should also check that retry report in task details, which I think has a bug (see prev comment)

Reply:
fixed that one, about task details, the `notification-send-skip-retry-still-report-failed-task-history-test` test should cover it, do you think we need to do it here too?

Reply:
That test just checks the schema, I think we need to test the data itself? Even if it's just the length.

Reply:
how about I add a length check in that schema?